src/pmhn/_trees/_backend_code.py

```diff
@@ -135,38 +45,75 @@ def loglikelihood(
+            lamb = np.exp(lamb)
+            subclone_lamb_map[i] = lamb
+        exit_lamb_map = {}
+        for i, (node, val) in enumerate(
+            list(tree_wrapper._genotype_subtree_node_map.keys())[-1]
+        ):
+            lineage = tree_wrapper._index_subclone_map[i]
+            lineage = list(lineage)
+            tree_mutations = set(lineage + [c.name for c in node.children])
+
+            exit_mutations = all_mut.difference(tree_mutations)
+
+            for mutation in exit_mutations:
+                lamb = 0
+                lamb += theta[mutation - 1][mutation - 1]
+                for j in lineage[1:]:
+                    lamb += theta[mutation - 1][j - 1]
+                lamb = np.exp(lamb)
+                exit_lamb_map[tuple(lineage + [mutation])] = lamb
+
+        V_old = np.zeros(subtrees_size)
+        V_old[0] = -1.0
+        V_new = np.zeros(subtrees_size)
+        x = 0.0
+        for genotype, index in tree_wrapper._genotype_subtree_node_map.items():
+            genotype_list = [item[1] for item in genotype]
+            for i, (node, val) in enumerate(genotype):
+                if val:
+                    lineage = tree_wrapper._index_subclone_map[i]
+                    lineage = list(lineage)
+                    tree_mutations = set(lineage + [c.name for c in node.children])
+
+                    exit_mutations = all_mut.difference(tree_mutations)
+
+                    for mutation in exit_mutations:
+                        subclone_index = tree_wrapper._subclone_index_map.get(
+                            tuple(lineage + [mutation])
                         )
-                    )
-                elif i == j:
-                    V_diag = sampling_rate - self._diag_entry(
-                        tree_wrapper, genotype_i, theta, all_mut
-                    )
-            for index, val in V_col:
-                x[i] -= val * x[index]
-            x[i] /= V_diag
-        return np.log(x[-1] + self._jitter) + np.log(sampling_rate)
+                        if subclone_index is None:
+                            lamb = exit_lamb_map[tuple(lineage + [mutation])]
+                            V_new[index] += lamb
+                        else:
+                            genotype_list[subclone_index] = 1
+                            ind = tree_wrapper._genotype_list_subtree_map.get(
+                                tuple(genotype_list)
+                            )
+
+                            lamb = subclone_lamb_map[subclone_index]
+                            V_new[ind] = -lamb
+                            V_new[index] += lamb
+                            genotype_list[subclone_index] = 0
+            V_new[index] += sampling_rate
+
+            x = -V_old[index] / V_new[index]
+            if index == subtrees_size - 1:
+                return np.log(x + self._jitter) + np.log(sampling_rate)
+            V_old += V_new * x
+            V_new = np.zeros_like(V_new)
+            x = 0.0
```

Comment:
Wouldn't it be clearer to initialize `x = 0.0` at the beginning of each iteration? Currently it's initialized before the loop starts and at the end of each iteration is zeroed again. (I could have overlooked something, though.)

Reply:
You're totally right, I'll change it.